Add support for rp2350 and dynamic flash sizing. Update to modern SDK. (extended)

tzstoyanov opened the pull request:
This patch updates the project to the current version of the SDK. It adds support for rp2350 and uses the flash size as set by the SDKs board file.

---
Based on @MelanieT's PR #5 with a few additional changes:
 - Addressed @JZimnol  comments .
 - Re-based on latest main branch.
 - Allow more flexible way for defining FOTA user options.
 - Updated for Mbed TLS 3.x
 
Tested on PicoW / RP2040.
Tested on Pico2 W / RP2350.

MelanieT committed:
Update app and boot linker scripts to new SDK version

MelanieT committed:
Add rp2350 support, add support for dynamic flash size.

Tzvetomir Stoyanov:
 - Use SDK defines to check for PICO platform in the C code.

MelanieT committed:
Allow reserving an area at the end of flash for a filesystem

tzstoyanov committed:
Allow redefine of FOTA compile time options

Usually FOTA is used as an external library to a project. There are
a few compile time options, that the user of the library can tweak
depending on the use case. Currently these can be changed only by
modifying the FOTA CMake file. Allowing user to define these options in
its CMake files simplifies the FOTA library integration.

tzstoyanov committed:
Fix for Mbed TLS 3.x

In Mbed TLS 3.x and later, some functions are renamed to match the new
API conventions. Added compatibility check and defines to handle both
old and new versions of the API.

tzstoyanov commented:
fixes #8

JZimnol commented on CMakeLists.txt:
```diff
@@ -21,15 +21,42 @@
 
 cmake_minimum_required(VERSION 3.13)
 
+if(${PICO_PLATFORM} STREQUAL "rp2350-arm-s")
```
```suggestion
if(${PICO_PLATFORM} STREQUAL "rp2350-arm-s" OR ${PICO_PLATFORM} STREQUAL "rp2350")
```

JZimnol commented on CMakeLists.txt:
```diff
@@ -21,15 +21,42 @@
 
+if(${PICO_PLATFORM} STREQUAL "rp2350-arm-s")
+    set(CPU "2350")
+else()
+    set(CPU "2040")
+endif()
```
```suggestion
elseif(${PICO_PLATFORM} STREQUAL "rp2040")
    set(CPU "2040")
else()
    message(FATAL_ERROR "Platform not supported: ${PICO_PLATFORM}")
endif()
```

JZimnol commented on CMakeLists.txt:
```diff
@@ -21,15 +21,42 @@
+    option(PFB_WITH_SHA256_HASHING "Enables image SHA256 appending and checking" ON)
+endif()
+if ( NOT PFB_RESERVED_FILESYSTEM_SIZE_KB )
+    set(PFB_RESERVED_FILESYSTEM_SIZE_KB 0)
```
```suggestion
    set(PFB_RESERVED_FILESYSTEM_SIZE_KB 0 CACHE STRING "EXPERIMENTAL: Size in KB reserved for the filesystem")
```

JZimnol commented on CMakeLists.txt:
```diff
@@ -21,15 +21,42 @@
+endif()
 
+########################################
+# Build linker definitions
+########################################
+# PICO_FLASH_SIZE_BYTES
+execute_process(
+    COMMAND cat ${CMAKE_CURRENT_LIST_DIR}/linker_common/linker_definitions.in
+    COMMAND ${CMAKE_C_COMPILER} -DFLASH_SIZE=${PICO_FLASH_SIZE_BYTES} -DFILESYSTEM_SIZE=${PFB_RESERVED_FILESYSTEM_SIZE_KB} -E -P -o ${CMAKE_BINARY_DIR}/linker_definitions.ld -
+    )
```
```suggestion
execute_process(
    COMMAND cat ${CMAKE_CURRENT_LIST_DIR}/linker_common/linker_definitions.in
    COMMAND ${CMAKE_C_COMPILER} -DFLASH_SIZE=${PICO_FLASH_SIZE_BYTES}
            -DFILESYSTEM_SIZE=${PFB_RESERVED_FILESYSTEM_SIZE_KB} -E -P
            -o ${CMAKE_BINARY_DIR}/linker_definitions.ld -)

```

JZimnol commented on CMakeLists.txt:
```diff
@@ -89,8 +116,14 @@ set(BOOTLOADER_DIR_GLOBAL ${CMAKE_CURRENT_SOURCE_DIR} PARENT_SCOPE)
+        set(CPU "2350")
+    else()
+        set(CPU "2040")
+    endif()
```
I think when you use `PARENT_SCOPE` in the same `if` above, you don't have to repeat it here.
Like:
```cmake
if(${PICO_PLATFORM} STREQUAL "rp2350-arm-s")
    set(CPU "2350" PARENT_SCOPE)
else()
    set(CPU "2040" PARENT_SCOPE)
endif()
```

tzstoyanov replied:
I've removed setting CPU in the function and made the other one with global scope.

JZimnol commented on linker_common/boot2350.ld:
```diff
@@ -0,0 +1,312 @@
+        __flash_info_should_rollback = .;
+        /* after flashing bootloader, rollback shouldn't be performed */
+        LONG(0x00000000)
+    } > FLASH_INFO
```
Don't you miss the whole 
```
ASSERT(__flash_info_app_vtor == __FLASH_INFO_APP_HEADER,
            "__FLASH_INFO_APP_HEADER definition in linker_definitions.ld file is not valid")
...
```
section as in `boot2040.ld`?

JZimnol commented on linker_common/linker_definitions.in:
```diff
@@ -45,6 +45,8 @@
 */
 
 __FLASH_START = 0x10000000;
+__FILESYSTEM_SIZE = (FILESYSTEM_SIZE * 1024);
+__FLASH_SIZE = FLASH_SIZE - __FILESYSTEM_SIZE;
```
I think the layout in ASCII (above and in the README.md) needs update too.

JZimnol commented on src/bootloader_app.c:
```diff
@@ -23,14 +23,22 @@
 #include <pico/stdlib.h>
 
 #include <pico_fota_bootloader/core.h>
+#include <stdlib.h>
```
`<pico/stdlib.h>` is already included above

MelanieT replied:
pico/stdlib.h is not the same as stdlib.h. I don't know if it's still needed but during development I needed something from the "real" stdlib.h

tzstoyanov replied:
Build passes without `#include <stdlib.h>`, I've removed it.

JZimnol commented on src/bootloader_app.c:
```diff
@@ -150,6 +159,7 @@ static void print_welcome_message(void) {
     puts("*                                                         *");
     puts("***********************************************************");
     puts("");
+    printf("Maximum code length: %luK\r\n", space);
```
```suggestion
    printf("[BOOTLOADER] Maximum code length: %luK\r\n", space);
```

JZimnol reviewed (changes requested):
Looks good! Please update also README's "Prerequisites" section and memory layout.

First round of review done. Unfortunately I cannot test it on pico2w as I don't have any.

tzstoyanov committed:
Address review comments

tzstoyanov committed:
Documentation update

Updated documentation with latest changes:
 - Support for Pico2 W.
 - File system block reservation.

JZimnol commented:
@tzstoyanov Please change the README.md prerequisites section from:
```md
## Prerequisites

- `pico-sdk` version `>= 1.5.1`

  - required for the `pico_mbedtls` library

- `Python 3` with the following packages: `argparse`, `hashlib`, `os`,
  `Crypto.Cipher`

  - required for the SHA256 calculation and AES ECB image encryption
```

to 

```md
## Prerequisites

- `pico-sdk` version `>= 2.2.0`

- `Python 3` with the following packages: `argparse`, `hashlib`, `os`,
  `Crypto.Cipher`

  - required for the SHA256 calculation and AES ECB image encryption
```

Otherwise I believe it looks good.

tzstoyanov commented:
> @tzstoyanov Please change the README.md prerequisites section from:
> 
> ```
> ## Prerequisites
> 
> - `pico-sdk` version `>= 1.5.1`
> 
>   - required for the `pico_mbedtls` library
> 
> - `Python 3` with the following packages: `argparse`, `hashlib`, `os`,
>   `Crypto.Cipher`
> 
>   - required for the SHA256 calculation and AES ECB image encryption
> ```
> 
> to
> 
> ```
> ## Prerequisites
> 
> - `pico-sdk` version `>= 2.2.0`
> 
> - `Python 3` with the following packages: `argparse`, `hashlib`, `os`,
>   `Crypto.Cipher`
> 
>   - required for the SHA256 calculation and AES ECB image encryption
> ```
> 
> Otherwise I believe it looks good.

Done, although I think it should work with SDK 1.5.1 (haven't tested with older SDK).
Thanks!
